File: Sean的leetcode/Leetcode/Guess Number Higher or Lower.cpp
/*
Binary Search
Time Complexity:O(logn)
*/
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        int l = 1;
        int r = n;
        while (l + 1 < r) {
            int mid = l + (r-l) / 2;
            int res = guess(mid);
            if (res == 0) {
                return mid;
            }
            else if (res == -1) {
                r = mid;
            }
            else {
                l = mid;
            }
        }
        return guess(l) == 0? l : r; 
    }
};

File: MyLeetCode/python/Guess Number Higher or Lower.py
###
# Binary Search
# Time Complexit: O(logn)
# Space Complexity: O(1)
###
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num):

class Solution(object):
    def guessNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n <= 1:
            return n
        l, r = 1, n
        while l + 1 < r:
            mid = (l + r) / 2
            g = guess(mid)
            if g == 0:
                return mid
            elif g < 0:
                r = mid
            else:
                l = mid
        return l if guess(l) == 0 else r
        